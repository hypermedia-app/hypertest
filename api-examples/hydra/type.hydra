With Operation <http://example.com/CreateProject> {
    Invoke {
        Expect Type <http://example.com/ProjectDraft>
    }
}
