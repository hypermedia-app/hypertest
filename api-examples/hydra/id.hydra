PREFIX ex: <http://example.com/>

With Operation ex:CreateProject {
    Invoke {
        Content-Type "text/turtle"

        ```
        @prefix schema: <http://schema.org/>

        <> schema:name "Testing with Hypertest" .
        ```
    } => {
        Expect Id <http://my.blog/article/testing-with-hypertest>
    }
}
