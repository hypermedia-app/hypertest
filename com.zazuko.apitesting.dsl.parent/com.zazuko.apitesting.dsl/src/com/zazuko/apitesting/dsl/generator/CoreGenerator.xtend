/*
 * generated by Xtext 2.17.0
 */
package com.zazuko.apitesting.dsl.generator

import com.zazuko.apitesting.dsl.core.ClassBlock
import com.zazuko.apitesting.dsl.core.ClassLevelAssertion
import com.zazuko.apitesting.dsl.core.PropertyAssertion
import org.eclipse.emf.ecore.resource.Resource
import org.eclipse.xtext.generator.AbstractGenerator
import org.eclipse.xtext.generator.IFileSystemAccess2
import org.eclipse.xtext.generator.IGeneratorContext

/**
 * Generates code from your model files on save.
 * 
 * See https://www.eclipse.org/Xtext/documentation/303_runtime_concepts.html#code-generation
 */
class CoreGenerator extends AbstractGenerator {

	override void doGenerate(Resource resource, IFileSystemAccess2 fsa, IGeneratorContext context) {
		
		val Iterable<ClassBlock> blocks = resource.allContents.filter(ClassBlock).toList
		
		if ( ! blocks.empty) {
			val String dslFileName = resource.getURI().lastSegment.toString();
			val String outFileBase = dslFileName.substring(0, dslFileName.lastIndexOf("."));
			
			fsa.generateFile(outFileBase + '.json', generateSteps(blocks));
		}		
	}	
	
	def generateSteps(Iterable<ClassBlock> blocks) '''
		{
			"steps": [
				«FOR block:blocks SEPARATOR ","»
					«block.step»
				«ENDFOR»
			]
		}
	'''
	
	def step(ClassBlock it) '''
		{
			"type": "Class",
			"classId": "«name»",
			"children": [
				«FOR assertion:assertions SEPARATOR ","»
					«assertion.child»
				«ENDFOR»
			]
			«hatch»
		}
	'''	
	
	def dispatch CharSequence child(PropertyAssertion it)  '''
		{
			"type": "Property",
			"propertyId": "«name»",
			"children": [
				«FOR assertion:assertions SEPARATOR ","»
					«assertion.child»
				«ENDFOR»
			]
			«hatch»
		}
	'''
	def dispatch child(ClassLevelAssertion it) '''
		# TODO: implementation missing for child(«class.name»)
	'''
	
}