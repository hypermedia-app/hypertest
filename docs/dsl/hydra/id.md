# Identifier Statement

`Expect Id` is a Hydra-specific expectation step similar to [`Expect Property`](../core/property)
which generates checks that a resource has the given identifier.

## Usage

<<< @/api-examples/hydra/id.hydra
