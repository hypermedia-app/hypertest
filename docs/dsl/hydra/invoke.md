# Invoke Block

Use the invoke block to define operation's request message and 

## Syntax

* `Invoke` keyword can be followed by one or two blocks, to set up request message and response steps 

## Usage

### Only check response message

<<< @/api-examples/hydra/operation/top-level.hydra

### Setting request headers

<<< @/api-examples/hydra/invoke/request-headers.hydra

### Sending request body

Request body can be set in two ways. One is to provide the contents inline,
delimited by triple backticks similarly to markdown.

:::warning
The parser preserves all whitspace except leading a trailing. If whitespace is
signinficant load the body from a file.
:::

<<< @/api-examples/hydra/invoke/request-inline-body.hydra

Second way is to reference a file whose contents will be sent.

<<< @/api-examples/hydra/invoke/request-file-body.hydra
