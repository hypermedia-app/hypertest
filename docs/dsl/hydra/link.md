# Link Step

Hydra extends the link's semantics by adding the possibility to link a URI template which has to be filled in by the client.

Hydra's DSL dialect facilitates by extending the syntax of the link step with a template variables section as a key/value set.

## Syntax

* Keys are URIs of prefixed names
* Values are string
* Same key can appear multiple times

## Usage

<<< @/api-examples/hydra/link/templated.hydra
