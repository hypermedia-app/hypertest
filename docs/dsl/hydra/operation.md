# Operation Step

Operation steps let the test authors perform requests to the server based on the
[`SupportedOperations`](http://www.hydra-cg.com/spec/latest/core/#example-9-defining-a-class-and-documenting-its-supported-operations)  
advertised by Hydra API Documentation or inlined in the resource representation. 

## Syntax

* Operation steps are valid where representation is the context.
* Operation steps are also valid on top level.
* Operation has a single child step [`Invoke`](./invoke).

## Usage

### Top-level scenario step

The simplest use of an operation is to place it in the root of a test scenario so that
it will be invoked as soon as any resource representation which supports it is analysed.

<<< @/api-examples/hydra/operation/top-level.hydra

### Context-specific operation

An operation can be limited in scope to a specific [`Class`](../core/class) or
anywhere else where representation steps are valid, such as [`Property`](../core/property).
Thus, the child invocations will only be performed when the operation is found within
a specific client state.

<<< @/api-examples/hydra/operation/context-specific.hydra

### Making operation mandatory

Just like with the [`Property`](../core/property) step, a missing operation can be
reported as a test failure by swapping `With` modifier for `Expect`.

<<< @/api-examples/hydra/operation/mandatory-operation.hydra

Child invocations can also be skipped to only check that the operation is
supported. If a top-level `With Operation` is found it will still be executed. 

<<< @/api-examples/hydra/operation/no-invocations.hydra
