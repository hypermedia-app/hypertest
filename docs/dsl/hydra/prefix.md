# Prefix Statement

Using prefixes it is possible to abbreviate class and operation identifiers by removing
the namespace URIs.

## Syntax

* Prefix step must appear before any other steps at the top of a scenario document

## Usage

<<< @/api-examples/hydra/prefix/prefixed-names.hydra
