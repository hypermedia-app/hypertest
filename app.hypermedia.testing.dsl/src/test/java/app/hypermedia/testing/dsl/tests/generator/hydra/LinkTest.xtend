package app.hypermedia.testing.dsl.tests.generator.hydra

import org.eclipse.xtext.generator.InMemoryFileSystemAccess
import org.junit.jupiter.api.^extension.ExtendWith
import org.eclipse.xtext.testing.extensions.InjectionExtension
import org.eclipse.xtext.testing.InjectWith
import com.google.inject.Inject
import org.eclipse.xtext.testing.util.ParseHelper
import app.hypermedia.testing.dsl.hydra.HydraScenario
import org.junit.jupiter.api.Test
import static io.github.jsonSnapshot.SnapshotMatcher.*
import org.eclipse.xtext.generator.IGenerator2
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.AfterAll
import org.json.JSONObject
import app.hypermedia.testing.dsl.tests.HydraInjectorProvider
import org.eclipse.xtext.generator.GeneratorContext

@ExtendWith(InjectionExtension)
@InjectWith(HydraInjectorProvider)
class LinkTest {
    @Inject IGenerator2 generator
    @Inject extension ParseHelper<HydraScenario>
    @BeforeAll
    static def beforeAll() {
        start()
    }

    @AfterAll
    static def afterAll() {
        validateSnapshots();
    }

    @Test
    def canGenerateForTemplatedLink() {
        // given
        val model = '''
            With Class <http://example.com/People> {
                With Link <http://example.com/search> (
                    <http://example.com/from> = "2010-12-12"
                    <http://example.com/to> = "2011-12-12"
                ) {
                    Expect Property <http://example.com/count> 5
                }
            }
        '''.parse

        // when
        val fsa = new InMemoryFileSystemAccess()
        generator.doGenerate(model.eResource, fsa, new GeneratorContext())
        println(fsa.textFiles)

        // then
        val file = new JSONObject(fsa.textFiles.values.get(0).toString)
        expect(file).toMatchSnapshot()
    }

    @Test
    def canGenerateForTemplatedLinkWithPrefix() {
        // given
        val model = '''
            PREFIX ex: <http://example.com/>

            With Class ex:People {
                With Link ex:search (
                    ex:from = "2010-12-12"
                    ex:to = "2011-12-12"
                ) {
                    Expect Property <http://example.com/count> 5
                }
            }
        '''.parse

        // when
        val fsa = new InMemoryFileSystemAccess()
        generator.doGenerate(model.eResource, fsa, new GeneratorContext())
        println(fsa.textFiles)

        // then
        val file = new JSONObject(fsa.textFiles.values.get(0).toString)
        expect(file).toMatchSnapshot()
    }
}
