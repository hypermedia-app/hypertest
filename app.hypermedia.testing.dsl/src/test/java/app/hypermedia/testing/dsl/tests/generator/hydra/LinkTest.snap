app.hypermedia.testing.dsl.tests.generator.hydra.LinkTest.canGenerateForTemplatedLink=[
  {
    "map": {
      "steps": {
        "myArrayList": [
          {
            "map": {
              "children": {
                "myArrayList": [
                  {
                    "map": {
                      "children": {
                        "myArrayList": [
                          {
                            "map": {
                              "propertyId": "http://example.com/count",
                              "strict": true,
                              "type": "Property",
                              "value": 5
                            },
                            "empty": false
                          }
                        ],
                        "empty": false
                      },
                      "rel": "http://example.com/search",
                      "strict": false,
                      "type": "Link",
                      "variables": {
                        "myArrayList": [
                          {
                            "map": {
                              "key": "http://example.com/from",
                              "value": "2010-12-12"
                            },
                            "empty": false
                          },
                          {
                            "map": {
                              "key": "http://example.com/to",
                              "value": "2011-12-12"
                            },
                            "empty": false
                          }
                        ],
                        "empty": false
                      }
                    },
                    "empty": false
                  }
                ],
                "empty": false
              },
              "classId": "http://example.com/People",
              "type": "Class"
            },
            "empty": false
          }
        ],
        "empty": false
      }
    },
    "empty": false
  }
]


app.hypermedia.testing.dsl.tests.generator.hydra.LinkTest.canGenerateForTemplatedLinkWithMultipleValuesForSameKey=[
  {
    "map": {
      "steps": {
        "myArrayList": [
          {
            "map": {
              "children": {
                "myArrayList": [
                  {
                    "map": {
                      "children": {
                        "myArrayList": [
                          {
                            "map": {
                              "propertyId": "http://example.com/count",
                              "strict": true,
                              "type": "Property",
                              "value": 5
                            },
                            "empty": false
                          }
                        ],
                        "empty": false
                      },
                      "rel": "http://example.com/search",
                      "strict": false,
                      "type": "Link",
                      "variables": {
                        "myArrayList": [
                          {
                            "map": {
                              "key": "http://example.com/name",
                              "value": "John"
                            },
                            "empty": false
                          },
                          {
                            "map": {
                              "key": "http://example.com/name",
                              "value": "Jane"
                            },
                            "empty": false
                          }
                        ],
                        "empty": false
                      }
                    },
                    "empty": false
                  }
                ],
                "empty": false
              },
              "classId": "http://example.com/People",
              "type": "Class"
            },
            "empty": false
          }
        ],
        "empty": false
      }
    },
    "empty": false
  }
]


app.hypermedia.testing.dsl.tests.generator.hydra.LinkTest.canGenerateForTemplatedLinkWithPrefix=[
  {
    "map": {
      "steps": {
        "myArrayList": [
          {
            "map": {
              "children": {
                "myArrayList": [
                  {
                    "map": {
                      "children": {
                        "myArrayList": [
                          {
                            "map": {
                              "propertyId": "http://example.com/count",
                              "strict": true,
                              "type": "Property",
                              "value": 5
                            },
                            "empty": false
                          }
                        ],
                        "empty": false
                      },
                      "rel": "http://example.com/search",
                      "strict": false,
                      "type": "Link",
                      "variables": {
                        "myArrayList": [
                          {
                            "map": {
                              "key": "http://example.com/from",
                              "value": "2010-12-12"
                            },
                            "empty": false
                          },
                          {
                            "map": {
                              "key": "http://example.com/to",
                              "value": "2011-12-12"
                            },
                            "empty": false
                          }
                        ],
                        "empty": false
                      }
                    },
                    "empty": false
                  }
                ],
                "empty": false
              },
              "classId": "http://example.com/People",
              "type": "Class"
            },
            "empty": false
          }
        ],
        "empty": false
      }
    },
    "empty": false
  }
]